demos/python_demos/handwritten_japanese_recognition_demo/utils/codec.py

```diff
@@ -20,26 +22,43 @@ def decode(self, preds):
         texts = []
         index = 0
         # Select max probabilty (greedy decoding) then decode index to character
-        preds_index = np.argmax(preds, 2)
-        preds_index = preds_index.transpose(1, 0)
-        preds_index_reshape = preds_index.reshape(-1)
-        preds_sizes = np.array([preds_index.shape[1]] * preds_index.shape[0])
-
-        for l in preds_sizes:
-            t = preds_index_reshape[index:index + l]
-
-            # NOTE: t might be zero size
-            if t.shape[0] == 0:
-                continue
-
-            char_list = []
-            for i in range(l):
-                # removing repeated characters and blank.
-                if t[i] != 0 and (not (i > 0 and t[i - 1] == t[i])):
-                    char_list.append(self.characters[t[i]])
-            text = ''.join(char_list)
-            texts.append(text)
-
-            index += l
+        preds_index = np.argmax(preds, 2) # WBD - > WB
+        preds_index = preds_index.transpose(1, 0) # WB -> BW
+        preds_index_reshape = preds_index.reshape(-1) # B*W
+
+        char_list = []
+        if self.designated_characters != None:
+            designated_character_list = []
+            with open(self.designated_characters, encoding='utf-8') as f:
+                lines = f.readlines()
+                for line in lines:
+                    designated_character_list.append(line.strip())
```

Comment:
You should move that reading to `__init__`. Otherwise the file is reread if `-ni` key > 1.

Reply:
Now `self.designated_characters` is not needed to be stored. In `__init()__` you can use `designated_characters` directly and in `decode()` you can check if `self.designated_character_list` is not None instead of `self.designated_characters`.